compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/PropertyReferenceLowering.kt

```diff
@@ -0,0 +1,350 @@
+import org.jetbrains.kotlin.ir.util.*
+import org.jetbrains.kotlin.ir.visitors.*
+import org.jetbrains.kotlin.load.java.JvmAbi
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.name.Name
+import org.jetbrains.kotlin.types.Variance
+
+internal val propertyReferencePhase = makeIrFilePhase(
+    ::PropertyReferenceLowering,
+    name = "PropertyReference",
+    description = "Construct KProperty instances returned by expressions such as A::x and A()::x"
+)
+
+internal class PropertyReferenceLowering(val context: JvmBackendContext) : ClassLoweringPass {
+    // TODO: join IrLocalDelegatedPropertyReference and IrPropertyReference via the class hierarchy?
+    private val IrMemberAccessExpression.getter: IrFunctionSymbol?
+        get() = (this as? IrPropertyReference)?.getter ?: (this as? IrLocalDelegatedPropertyReference)?.getter
+
+    private val IrMemberAccessExpression.setter: IrFunctionSymbol?
+        get() = (this as? IrPropertyReference)?.setter ?: (this as? IrLocalDelegatedPropertyReference)?.setter
+
+    private val IrMemberAccessExpression.field: IrFieldSymbol?
+        get() = (this as? IrPropertyReference)?.field
+
+    // Reflection metadata for local properties is serialized under the signature "<v#$N>" attached to the containing class.
+    // This maps properties to values of N.
+    private val localPropertyIndices = mutableMapOf<IrSymbol, Int>()
+
+    private val IrMemberAccessExpression.signature: String
+        get() = localPropertyIndices[getter]?.let { "<v#$it>" }
+            ?: getter?.owner?.let { context.state.typeMapper.mapSignatureSkipGeneric(it.descriptor).toString() }
+            // Plain Java fields do not have a getter, but can be referenced nonetheless.
+            ?: TODO("plain Java field signature")
+
+    private val IrMemberAccessExpression.symbol: IrSymbol
+        get() = getter?.owner?.symbol ?: field!!.owner.symbol
+
+    private val plainJavaClass =
+        context.getIrClass(FqName("java.lang.Class")).owner
+
+    private val reflectionClass =
+        context.getIrClass(FqName("kotlin.jvm.internal.Reflection")).owner
+
+    private val getOrCreateKotlinClass =
+        reflectionClass.functions.single { it.name.asString() == "getOrCreateKotlinClass" && it.valueParameters.size == 1 }
+
+    private val getOrCreateKotlinPackage =
+        reflectionClass.functions.single { it.name.asString() == "getOrCreateKotlinPackage" && it.valueParameters.size == 2 }
+
+    private val arrayItemGetter =
+        context.ir.symbols.array.owner.functions.single { it.name.asString() == "get" }
+
+    private val kPropertyType =
+        context.getIrClass(FqName("kotlin.reflect.KProperty")).owner.defaultType
+
+    private val kPropertiesFieldType =
+        context.ir.symbols.array.createType(false, listOf(makeTypeProjection(kPropertyType, Variance.OUT_VARIANCE)))
+
+    // Return some declaration the parent of which is the containing class/package for the referenced property.
```

Comment:
Maybe best to move this lowering after LocalDeclarationsLowering?

Reply:
Right now I've placed it before PropertiesToFieldsLowering because otherwise support for Java fields would also be a requirement for some Kotlin properties. Once that's implemented, maybe there's a more optimal place in the pipeline.

Reply:
Added a TODO about that.